pkg/kubelet/nodestatus/setters.go

```diff
@@ -88,6 +88,12 @@ func NodeAddress(nodeIP net.IP, // typically Kubelet.nodeIP
 			if err != nil {
 				return err
 			}
+			// For every address supplied by the cloud provider that matches nodeIP, nodeIP is the enforced node address for
+			// that address Type (like InternalIP and ExternalIP), meaning other addresses of the same Type are discarded.
+			// This ensures there is one address of that Type on the node that corresponds to nodeIP.
+			// See #61921 for more information: some cloud providers may supply secondary IPs, and it's assumed elsewhere
```

Comment:
Thanks for the comment, it helps clear things up.  Remind me why we don't fix cloudproviders/the kubelet interface to only provide one address of each address type, since it seems that's what the API server is expecting?  Why do we have this convoluted logic to accept multiple addresses of each type, narrow them down to one based on the value of node IP, etc?

Reply:
My best guess is that the interface/contract for `nodeAddressesFunc` was never well-defined and so individual cloudproviders have been band-aiding fixes to accommodate their implementations. For example, see this comment from 4 years ago:
`// TODO: Other IP addresses (multiple ips)?` https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/legacy-cloud-providers/aws/aws.go#L1506

Reply:
@nckturner Why?  The whole concept of node addresses was never defined - exactly what semantics were implied/desired by internalIPs vs externalIPs, etc.  A long succession of arbitrary local decisions and terminologies were merged to deal with specific cloud provider's unique networking and billing situations.  To compound it, all this gets ignored on GCE (where hostname == DNS == unique provider id == k8s node name) so there were no tests enforcing other combinations until ~recently and GCE-privileged reviewers who didn't even understand that there were other legitimate combinations.  (Slightly bitter, since I've been fighting/losing this battle for many years.)

Afaics there was never a reason for more than one node IP address (not even one per address family), other than exposing debugging information to humans (which is important, but different).

Reply:
I don't think it's true that "it's assumed elsewhere that there is only one address of each Type on a node".

There is definitely an assumption that the *first* InternalIP is the IP that other nodes and masters should use to reach the node, but I'm not aware of any code that assumes there won't be any other InternalIPs. (Well... I suppose the NodeStatus strategic merge annotation assumed that, but it's not used any more because that assumption was wrong.)

Historically the node has reported additional IPs that the cloud provider reports, because presumably there is some reason the cloud provider is reporting them, and thus the user/administrator might want to see them.

Reply:
Removed, thank you for the explanation!